### PR TITLE
Add Testdata in json and implement a test runner

### PR DESCRIPTION
**Pull request by KilnOfTheSecondFlame**

### CODE REVIEW

**Comment by KilnOfTheSecondFlame** [comment]
I assume an integration test is failing:
grafik

File: src/main/kotlin/ApplicationStarter.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import kotlin.random.Random
+
+fun main() {
+    val scannerAbstractionImpl = ScannerAbstractionImpl()
+    val guiLess = GuiLess(scannerAbstractionImpl)
+    val jsonReadableFile = ReadableResourceFile("package.json")
+    val dataLoader = JsonDataLoader(jsonReadableFile)
+    val txtFile = File("./Report.txt")
+    val reportPersistor = ReportTxtPersistor(WriteableFileImpl(txtFile))
+    val random = Random(1337)
```
Comment (CryDeb) [inline src/main/kotlin/ApplicationStarter.kt:19]: Random mit seed = statischer wert

File: src/main/kotlin/MultipleSequentialTestExecutor.kt [hunk 2]
```diff
@@ -1 +1,244 @@
...
+                "`----------------------------------------------------------------´;;;;;;;;;;;;;;;;;;;`     \n"
+        guiLess.display(header)
+    }
+
+    fun start() {
+        while (true) {
+            displayHeader()
+            guiLess.display("---- Tests ----")
+            testDataList.iterator().forEach { testData -> guiLess.display("${testData.id} - ${testData.name}") }
+            guiLess.display("")
```
Comment (CryDeb) [inline src/main/kotlin/MultipleSequentialTestExecutor.kt:66]: ist das ein lineFeed? -> falls ja ist eher unschön

File: src/main/kotlin/MultipleSequentialTestExecutor.kt [hunk 3]
```diff
@@ -1 +1,244 @@
...
+            )
+        }
+    }
+
+    private fun quit() {
+        saveReportWithoutAsking()
+        System.exit(0)
+    }
+
+    private fun executeTest(testData: TestData) {
```
Comment (CryDeb) [inline src/main/kotlin/MultipleSequentialTestExecutor.kt:80]: Viel zu lang -> 
Schaue dass funktionen klein bleiben.

Genaugenommen machst du einen Zustandsautomaten -> da könntest du sicher eine klasse oder zwei dafür bauen.

File: src/main/kotlin/MultipleSequentialTestExecutor.kt [hunk 4]
```diff
@@ -1 +1,244 @@
...
+        var tagInformation: TagInformation
+        while (tagIdRequested) {
+            guiLess.getUserInput("Tag ID")
+        }
+        tagInformation = TagInformation(tagId)
+        return tagInformation
+    }
+
+    override fun handleUserInput(userMessage: String) {
+        try {
```
Comment (CryDeb) [inline src/main/kotlin/MultipleSequentialTestExecutor.kt:207]: Diese Funktion gefällt mir weniger gut -> viele verschachtelung und grosse funktion.
Könnte auch hier eine eigene Klasse sin ergeben?

File: src/main/kotlin/test/cases/runner/MultipleTagsMultipleReads.kt [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import rfid.communication.TagInformation
+import kotlin.system.measureTimeMillis
+
+class MultipleTagsMultipleReads(communicationDriver: CommunicationDriver) : TestRunner(communicationDriver) {
+    override fun run() {
+        testResults = ""
+        for (i in 1..amountOfReads) {
+            val tagsInVicinity = mutableListOf<TagInformation>()
+            val timeForRead = measureTimeMillis { tagsInVicinity.addAll(communicationDriver.getAllRfids(10)) }
+            // TODO what happens when timeout is reached? Has to be handled
```
Comment (CryDeb) [inline src/main/kotlin/test/cases/runner/MultipleTagsMultipleReads.kt]: !

File: src/main/kotlin/test/cases/runner/MultipleTagsSingleRead.kt [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+import rfid.communication.CommunicationDriver
+import rfid.communication.TagInformation
+import kotlin.system.measureTimeMillis
+
+class MultipleTagsSingleRead(communicationDriver: CommunicationDriver) : TestRunner(communicationDriver) {
+    override fun run() {
+        testResults = ""
+        val tagsInVicinity = mutableListOf<TagInformation>()
+        val timeForRead = measureTimeMillis { tagsInVicinity.addAll(communicationDriver.getAllRfids(10)) }
+        // TODO what happens when timeout is reached? Has to be handled
```
Comment (CryDeb) [inline src/main/kotlin/test/cases/runner/MultipleTagsSingleRead.kt]: !!

File: src/main/kotlin/test/cases/runner/TestRunner.kt [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package test.cases.runner
+
+import rfid.communication.CommunicationDriver
+import rfid.communication.TagInformation
+
+abstract class TestRunner(val communicationDriver: CommunicationDriver) : Runnable {
```
Comment (CryDeb) [inline src/main/kotlin/test/cases/runner/TestRunner.kt:6]: ist interface möglich?

File: src/main/resources/package.json [hunk 8]
```diff
@@ -1,24 +1,158 @@
...
   {
     "id": 2,
-    "name": "TestName",
-    "testType": "A",
+    "name": "Lesegeschwindigkeit Bulk Reading",
+    "testType": "MultipleTagsMultipleReads",
     "preParameters": [
-      "Please insert some params"
+      "AmountOfReads",
+      "DistanceBetweenTagsAndAntenna",
```
Comment (CryDeb) [inline src/main/resources/package.json]: Jeder Text hier sollte wirklich leserlich sein, also nicht camelcase verwenden und allenfals auch besser in deutsch

File: src/test/kotlin/gui/GuiLessTestIntegration.kt [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        }
+    }
+
+    @Test
+    fun displayMultiple() {
+        val guiLess = GuiLess(scannerAbstractionImpl)
+        val messages = listOf("Element A", "Element B", "Element C")
+        guiLess.displayMultiple(messages)
+        messages.iterator().forEach { message ->
+            MatcherAssert.assertThat(
```
Comment (CryDeb) [inline src/test/kotlin/gui/GuiLessTestIntegration.kt]: schaue dass nur assertThat dargestellt wird.

File: src/test/kotlin/gui/GuiLessTestIntegration.kt [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+                resultStream.toString(),
+                CoreMatchers.containsString(message)
```
Comment (CryDeb) [inline src/test/kotlin/gui/GuiLessTestIntegration.kt]: siehe oben

**Review by CryDeb — CHANGES_REQUESTED** [review]

File: src/main/kotlin/ApplicationStarter.kt (hunk 1 above)
Comment (KilnOfTheSecondFlame) [reply src/main/kotlin/ApplicationStarter.kt:19]: Aber es ist ein Simulator, darum interessiert es uns nicht 😄

File: src/main/kotlin/MultipleSequentialTestExecutor.kt (hunk 2 above)
Comment (KilnOfTheSecondFlame) [reply src/main/kotlin/MultipleSequentialTestExecutor.kt:66]: Ja und nein, es ist eine Anweisung die Versuchsnummer und -Name auf der Zeile anzuzeigen

File: src/main/kotlin/test/cases/runner/TestRunner.kt (hunk 7 above)
Comment (KilnOfTheSecondFlame) [reply src/main/kotlin/test/cases/runner/TestRunner.kt:6]: Nein, da wir bei gewissen Versuchen die Tag ID erwarten

File: src/main/resources/package.json (hunk 8 above)
Comment (KilnOfTheSecondFlame) [reply src/main/resources/package.json]: 👍

File: src/main/kotlin/MultipleSequentialTestExecutor.kt [hunk 11]
```diff
@@ -29,30 +63,181 @@ class MultipleSequentialTestExecutor(
...
+            guiLess.display("Save as $filename")
+            guiLess.getUserInput("(y)es or write your own")
+        } else {
+            guiLess.display("\nThere is no report to save!\n")
+        }
+    }
+
+    private fun saveReportWithoutAsking() {
+        reportPersistor.changeFilename(filename)
+        if (lastReport != null) reportPersistor.persistReport(lastReport!!)
```
Comment (CryDeb) [inline src/main/kotlin/MultipleSequentialTestExecutor.kt:190]: lastReport?.let(reportPersistor::persistReport)

File: src/main/kotlin/feature/report/ReportTxtPersistor.kt [hunk 12]
```diff
@@ -1,8 +1,20 @@
...
+import java.io.File
 
 class ReportTxtPersistor(var file: WriteableFile) : ReportPersistor {
+    override fun changeFilename(filename: String) {
+        val filenameIsValid = FilenameValidator().isFilenameValid(filename)
+        if (filenameIsValid) {
+            file.changeFile(File("$filename.txt"))
+        }
+        else {
+            println("Filename is invalid")
```
Comment (CryDeb) [inline src/main/kotlin/feature/report/ReportTxtPersistor.kt:14]: throw exception

File: src/main/kotlin/feature/report/ReportTxtPersistor.kt [hunk 13]
```diff
@@ -1,8 +1,20 @@
 package feature.report
 
+import util.FilenameValidator
 import util.WriteableFile
+import java.io.File
 
 class ReportTxtPersistor(var file: WriteableFile) : ReportPersistor {
+    override fun changeFilename(filename: String) {
+        val filenameIsValid = FilenameValidator().isFilenameValid(filename)
```
Comment (CryDeb) [inline src/main/kotlin/feature/report/ReportTxtPersistor.kt:9]: Sollte in WritableFile sein

File: src/main/kotlin/test/cases/runner/MultipleTagsMultipleReads.kt (hunk 5 above)
Comment (CryDeb) [reply src/main/kotlin/test/cases/runner/MultipleTagsMultipleReads.kt]: !!

File: src/main/kotlin/test/cases/runner/MultipleTagsSingleRead.kt (hunk 6 above)
Comment (CryDeb) [reply src/main/kotlin/test/cases/runner/MultipleTagsSingleRead.kt]: !!!
